mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/transformation/mpsToModelix/initial/ModuleSynchronizer.kt

```diff
@@ -43,44 +45,57 @@ import org.modelix.mps.sync.transformation.cache.MpsToModelixMap
 import org.modelix.mps.sync.util.bindTo
 import org.modelix.mps.sync.util.nodeIdAsLong
 import org.modelix.mps.sync.util.waitForCompletionOfEachTask
+import java.util.Collections
 import java.util.concurrent.CompletableFuture
 
 @UnstableModelixFeature(reason = "The new modelix MPS plugin is under construction", intendedFinalization = "2024.1")
 class ModuleSynchronizer(private val branch: IBranch) {
 
+    private val logger = KotlinLogging.logger {}
     private val nodeMap = MpsToModelixMap
     private val syncQueue = SyncQueue
     private val bindingsRegistry = BindingsRegistry
 
     private val modelSynchronizer = ModelSynchronizer(branch, postponeReferenceResolution = true)
 
-    fun addModuleAndActivate(module: AbstractModule) {
-        addModule(module, true).continueWith(linkedSetOf(SyncLock.NONE), SyncDirection.NONE) {
-            @Suppress("UNCHECKED_CAST")
-            (it as? Iterable<IBinding>)?.forEach(IBinding::activate)
-        }
-    }
-
-    private fun addModule(
+    fun addModule(
         module: AbstractModule,
         isTransformationStartingModule: Boolean = false,
     ): ContinuableSyncTask =
         syncQueue.enqueue(linkedSetOf(SyncLock.MODELIX_WRITE, SyncLock.MPS_READ), SyncDirection.MPS_TO_MODELIX) {
             val rootNode = branch.getRootNode()
             val childLink = ChildLinkFromName("modules")
-            val concept = BuiltinLanguages.MPSRepositoryConcepts.Module
-            val cloudModule = rootNode.addNewChild(childLink, -1, concept)
 
-            nodeMap.put(module, cloudModule.nodeIdAsLong())
+            // duplicate check
+            val moduleId = module.moduleId.toString()
+            val moduleExists = rootNode.getChildren(childLink)
+                .firstOrNull { moduleId == it.getPropertyValue(BuiltinLanguages.MPSRepositoryConcepts.Module.id) } != null
+            if (moduleExists) {
+                if (nodeMap.isMappedToModelix(module)) {
+                    return@enqueue ItemAlreadySynchronizer(module)
+                } else {
+                    throw Exception("Module ${module.moduleName} already exists on the server, therefore it will not be synchronized. Remove it from the project and synchronize it from the server instead.")
+                }
+            }
 
+            val cloudModule = rootNode.addNewChild(childLink, -1, BuiltinLanguages.MPSRepositoryConcepts.Module)
+            nodeMap.put(module, cloudModule.nodeIdAsLong())
             synchronizeModuleProperties(cloudModule, module)
 
             // synchronize dependencies
             module.declaredDependencies.waitForCompletionOfEachTask(collectResults = true) { addDependency(module, it) }
         }.continueWith(linkedSetOf(SyncLock.NONE), SyncDirection.NONE) { unflattenedBindings ->
-            @Suppress("UNCHECKED_CAST")
-            (unflattenedBindings as Iterable<Iterable<IBinding>>).flatten()
+            if (unflattenedBindings is Iterable<*>) {
```

Comment:
Name `unflattenedBindings` is missleading because it also can be an `ItemAlreadySynchronizer`.